sdks/python/apache_beam/internal/pickler.py

```diff
@@ -189,11 +189,22 @@ def dumps(o):
     s = dill.dumps(o)
   finally:
     dill.dill._trace(False)  # pylint: disable=protected-access
-  return base64.b64encode(zlib.compress(s))
+
+  # Compress as compactly as possible to decrease peak memory usage (of multiple
+  # in-memory copies). Also use a separate statement in order to decrease peak
+  # memory usage (by allowing unneeded data to go out of scope).
+  s = zlib.compress(s, 9)
```

Comment:
Done (using the del suggestion).

In principle this could be done in a streaming manner, but I think for both cases the majority of the memory is the uncompressed object (and similarly for the b64 stuff).

Furthermore, according to the comments at the top of the file this should be used for functions and modules (as opposed to data) so presumably overheads are not ridiculously large?